docs/style.md

```diff
@@ -109,6 +109,44 @@ Note that for some *big* `Copy` types, notably `CryptoHash`, we sometimes use
 references for performance reasons. As a rule of thumb, `T` is considered *big* if
 `size_of::<T>() > 2 * size_of::<usize>()`.
 
+### Prefer for loops over `for_each` and `try_for_each` methods
+
+Iterators offer `for_each` and `try_for_each` methods which allow executing
+a closure over all elements of the iterator.  This is similar to using a for
+loop but comes with various complications and may lead to less readable code.
+Prefer using a loop rather than those methods, for example:
+
+```rust
+// GOOD
+for outcome_with_id in result? {
+    *total_gas_burnt =
+        safe_add_gas(*total_gas_burnt, outcome_with_id.outcome.gas_burnt)?;
+    outcomes.push(outcome_with_id);
+}
+
+// BAD
+result?.into_iter().try_for_each(
+    |outcome_with_id: ExecutionOutcomeWithId| -> Result<(), RuntimeError> {
+        *total_gas_burnt =
+            safe_add_gas(*total_gas_burnt, outcome_with_id.outcome.gas_burnt)?;
+        outcomes.push(outcome_with_id);
+        Ok(())
+    },
+)?;
+```
+
+**Rationale:** The `for_each` and `try_for_each` method don’t play nice with
+`break` and `continue` statements or with async IO (since `.await` is not
+possible).  And while `try_for_each` allows for the use of question mark
+operator, one may end up having to uses it twice: once inside the closure and
+second time outside the call to `try_for_each`.  Furthermore, usage of the
+functions often introduce some minor syntax noise.
+
+However, there are situations when those methods may lead to more readable code.
```

Comment:
Not part of this PR and a question for @matklad as well.  Every rule has an exception and in certain cases, if someone demonstrates the need to break one of these rules then they should be allowed to make exceptions.  Do we have some style agreed upon on how to do this?

Reply:
I would say that styleguide is more of a lighthouse pointing the direction to go, rather than a set of rules that guard PRs. 

The following algo works:

**A** writes code with `.for_each`
**B** reviews and points to the style giude

Then, one of the three outcomes are possible. 

1. **A** changes to `for` because they don't care. 
2. **A** says "I aknowledge the styleguide, but in this particular cases there's this specific reason to prefer `.for_each`". `.for_each` stays without much arguming about specific reason
3. **A** says "but `.for_each` are more readable in general", in which case the discussion is moved out of scope for the particular PR. 



I think there should be a very low bar for making exceptions.

Reply:
I'm happy with this "process".